File: core/src/rustyice/game/actors/PointLight.java
package rustyice.game.actors;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.g2d.Sprite;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.graphics.g2d.TextureAtlas;
import rustyice.core.Core;
import rustyice.editor.annotations.ComponentAccess;
import rustyice.game.Actor;
import rustyice.game.lights.LightComponent;
import rustyice.game.lights.LightContainer;
import rustyice.game.lights.PointLightComponent;
import rustyice.game.physics.PointPhysicsComponent;
import rustyice.graphics.Camera;
import rustyice.graphics.RenderFlags;



public class PointLight extends Actor implements LightContainer {
    private transient Sprite icon;
    private PointLightComponent light;
    
    public PointLight() {
        light = new PointLightComponent();
        setPhysicsComponent(new PointPhysicsComponent());
        setSize(1, 1);
    }

    @Override
    public void render(SpriteBatch batch, Camera camera, int flags) {
        if((flags & RenderFlags.EDITOR) == RenderFlags.EDITOR){
            if(icon == null){
                icon = Core.resources.get("art.atlas", TextureAtlas.class).createSprite("EditorLight");
            }
            icon.setBounds(getX() - getWidth()/2, getY() - getHeight()/2, getWidth(), getHeight());
            icon.setOriginCenter();
            icon.setRotation(getRotation());
            icon.setColor(Color.YELLOW);
            icon.draw(batch);
        }
    }

    @Override
    public void init() {
        super.init();
        light.init(this);
        light.setPosition(getX(), getY());
    }

    @Override
    public void update(float delta) {
        super.update(delta);

    }
    
    @Override
    public void setX(float x) {
        super.setX(x);
        light.setPosition(x, getY());
    }
    
    @Override
    public void setY(float y) {
        super.setY(y);
        light.setPosition(getX(), y);
    }

    @Override
    public void setPosition(float x, float y) {
        super.setPosition(x, y);
        light.setPosition(x, y);
    }
    
    @Override
    public void store() {
        super.store();
        light.store();
    }

    @Override
    @ComponentAccess
    public LightComponent getLightComponent() {
        return light;
    }
}

File: core/src/rustyice/game/actors/PointLight.kt
package rustyice.game.actors

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.g2d.Batch
import com.badlogic.gdx.graphics.g2d.Sprite
import rustyengine.RustyEngine
import rustyice.editor.annotations.ComponentAccess
import rustyice.game.Actor
import rustyice.game.lights.LightComponent
import rustyice.game.lights.PointLightComponent
import rustyice.graphics.Camera
import rustyice.graphics.EDITOR
import rustyengine.resources.Resources


/**
 * @author gabek
 */
class PointLight() : Actor(), LightComponent.LightContainer {
    @Transient private var icon: Sprite? = null
    @ComponentAccess
    override val lightComponent = PointLightComponent()

    override var x: Float
        get() = super.x
        set(value) {
            super.x = value
            lightComponent.x = value
        }

    override var y: Float
        get() = super.y
        set(value) {
            super.y = value
            lightComponent.y = value
        }

    override fun render(batch: Batch, camera: Camera, renderFlags: Int) {
        lightComponent.render(batch, camera, renderFlags)

        if ((renderFlags or EDITOR) == EDITOR) {
            if (icon == null) {
                icon = RustyEngine.resorces.gameArt.createSprite("EditorLight")
            }

            icon?.let {
                it.setBounds(x - width / 2, y - height / 2, width, height)
                it.setOriginCenter()
                it.rotation = rotation
                it.color = Color.YELLOW
                it.draw(batch)
            }
        }
    }

    override fun init() {
        super.init()
        lightComponent.setPosition(x, y)
        lightComponent.parent = this
        lightComponent.init()
    }

    override fun update(delta: Float) {
        super.update(delta)
        lightComponent.update(delta)
    }

    override fun store() {
        super.store()
        lightComponent.store()
    }

    init {
        setSize(1f, 1f)
    }
}
